src/main/java/com/devcycle/sdk/server/local/LocalBucketing.java

```diff
@@ -0,0 +1,149 @@
+        linker.module(store, "", module); // linker now has a runnable instance of the module
+
+        Memory mem = linker.get(store, "", "memory").get().memory();
+        memRef.set(mem);
+    }
+
+    private Collection<Extern> setImportsOnLinker() {
+        Func dateNowFn = WasmFunctions.wrap(store, F64, () -> {
+            return (double) System.currentTimeMillis();
+        });
+        linker.define("env", "Date.now", Extern.fromFunc(dateNowFn));
+
+        Func consoleLogFn = WasmFunctions.wrap(store, I32, (addr) -> {
+            String message = readWasmString(((Number) addr).intValue());
+            System.out.println("Local Bucketing Log - " + message);
+        });
+        linker.define("env", "console.log", Extern.fromFunc(consoleLogFn));
+
+        Func abortFn = WasmFunctions.wrap(store, I32, I32, I32, I32, (messagePtr, filenamePtr, linenum, colnum) -> {
+            String message = readWasmString(((Number) messagePtr).intValue());
+            String fileName = readWasmString(((Number) filenamePtr).intValue());
+            throw new RuntimeException("Local Bucketing Module Aborted. File " + fileName + ", Line " + linenum + ":" + colnum + " \n Message: " + message + "\n");
+        });
+        linker.define("env", "abort", Extern.fromFunc(abortFn));
+
+        return Arrays.asList(Extern.fromFunc(dateNowFn), Extern.fromFunc(consoleLogFn), Extern.fromFunc(abortFn));
+    }
+
+    private int writeWasmString(String param) {
+
+        int objectIdString = 1; // id 1 represents string class in wasm
+
+        Func __newPtr = linker.get(store, "", "__new").get().func(); // get pointer to __new function
+        WasmFunctions.Function2<Integer, Integer, Integer> __new = WasmFunctions.func(
+                store, __newPtr, I32, I32, I32); // load __new function
+
+        byte[] paramBytes = param.getBytes(StandardCharsets.UTF_8);
+        int paramAddress = __new.call(paramBytes.length * 2, objectIdString); // allocate memory in store for a string with this length and get start address
+
+        ByteBuffer buf = memRef.get().buffer(store);
+        for (int i = 0; i < paramBytes.length; i++) {
+            buf.put(paramAddress + (i * 2), paramBytes[i]); // write each byte of string starting at address
+        }
+
+        return paramAddress;
+    }
+
+    private String readWasmString(int startAddress) {
+        ByteBuffer buf = memRef.get().buffer(store);
+
+        // objects in wasm memory have a 20 byte header before the start pointer
+        // the 4 bytes right before the object pointer store the length of the object as an unsigned int
+        // see assemblyscript.org/runtime.html#memory-layout
+        byte[] headerBytes = {buf.get(startAddress - 1), buf.get(startAddress - 2), buf.get(startAddress - 3), buf.get(startAddress - 4)};
+        long stringLength = getUnsignedInt(headerBytes);
+        String result = "";
+        for (int i = 0; i < stringLength; i += 2) { // +=2 because every 2nd byte is null
```

Comment:
Technically this comment is correct, but it's because the data is formatted as WTF-16 - not UTF-8. So technically the second byte is a valid data byte - but we're not using it, so we skip. Not that the byte is `\0`.

Reply:
oh that makes sense, i'll just put that it's formatted as WTF-16 instead of UTF-8, hopefully that will be more helpful